dropshot/src/pagination.rs

```diff
@@ -0,0 +1,569 @@
+ *
+ * We impose a maximum length primarily to prevent a client from making us parse
+ * extremely large strings.  We apply this limit when we create tokens to avoid
+ * handing out a token that can't be used.
+ *
+ * Note that these tokens are passed in the HTTP request line (before the
+ * headers), and many HTTP implementations impose an implicit limit as low as
+ * 8KiB on the size of the request line and headers together, so it's a good
+ * idea to keep this as small as we can.
+ */
+const MAX_TOKEN_LENGTH: usize = 512;
+
+/**
+ * Version for the pagination token serialization format
+ *
+ * This may seem like overkill, but it allows us to rev this in a future version
+ * of Dropshot without breaking any ongoing scans when the change is deployed.
+ * If we rev this, we might need to provide a way for clients to request at
+ * runtime which version of token to generate so that if they do a rolling
+ * upgrade of multiple instances, they can configure the instances to generate
+ * v1 tokens until the rollout is complete, then switch on the new token
+ * version.  Obviously, it would be better to avoid revving this version if
+ * possible!
+ *
+ * Note that consumers still need to consider compatibility if they change their
+ * own `ScanMode` or `PageSelector` types.
+ */
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    Deserialize,
+    ExtractedParameter,
+    JsonSchema,
+    PartialEq,
+    Serialize,
+)]
+#[serde(rename_all = "lowercase")]
+enum PaginationVersion {
+    V1,
+}
+
+/**
+ * Parts of the pagination token that actually get serialized.
+ */
+#[derive(Debug, Deserialize, Serialize)]
+struct SerializedToken<PageSelector> {
+    v: PaginationVersion,
+    page_start: PageSelector,
+}
+
+/**
+ * Describes the current scan and the client's position within it
+ */
+#[derive(Debug, Deserialize, ExtractedParameter)]
+#[serde(try_from = "String")]
+#[serde(bound(deserialize = "PageSelector: DeserializeOwned"))]
+pub struct PageToken<PageSelector> {
+    pub page_start: PageSelector,
+}
+
+impl<PageSelector> PageToken<PageSelector>
+where
+    PageSelector: Serialize,
+{
+    pub fn new(page_start: PageSelector) -> Self {
+        PageToken {
+            page_start,
+        }
+    }
+
+    pub(crate) fn to_serialized(self) -> Result<String, HttpError> {
+        let page_start = self.page_start;
+
+        let marker_bytes = {
+            let serialized_marker = SerializedToken {
+                v: PaginationVersion::V1,
+                page_start: page_start,
+            };
+
+            let json_bytes =
+                serde_json::to_vec(&serialized_marker).map_err(|e| {
+                    HttpError::for_internal_error(format!(
+                        "failed to serialize marker: {}",
+                        e
+                    ))
+                })?;
+
+            base64::encode_config(json_bytes, URL_SAFE)
+        };
+
+        /*
+         * TODO-robustness is there a way for us to know at compile-time that
```

Comment:
hah.. this TODO is exactly the question I was going to ask. How would we feel about just applying a SHA hash to larger tokens automatically? Is that too cute?

Reply:
I don't follow how we'd use that (because there's no way to reverse the SHA hash on the way back in, right?)

Reply:
Good point; I didn't think it through.